feat(menus): custom menus updates

kmcfaul opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Closes #8899

- Renames demos to not use the word 'composable'
- Fixes drilldown demo. The if block controlling height updates was no longer being reached because the initial heights being returned for non-visible menus was updated to be non-zero (because of padding). The issue mentions flyout also needed to be fixes but it looks like its working. Maybe it was fixed as byproduct of other updates, lmk if anything seems broken with it.
- Removes redundant demos (basic select and dropdown demos that are already covered in their respective docs)
- Converts the remaining demos to not use Popper. Demos now use Select or Dropdown, except for the flyout, drilldown, and tree view demos. Flyout and drilldown make extensive use of extra menu props not explicitly exposed by Select or Dropdown (though it would still work), as well as needing props on `MenuContent` which isn't accessible by Select/Dropdown currently (for drilldown). The tree view demo doesn't use menu at all. To keep from using Popper directly still, I added a `MenuContainer` component to menu to act as a similar wrapper for Popper like Select/Dropdown. The application launcher demo and context selector demo end up being slightly different in the DOM than previously because `MenuContent` is internal to Select/Dropdown and both had elements (MenuSearch, MenuFooter) that previously were outside of the content. I also had to add an `onOpenChangeKeys` prop to Dropdown because these two require the `Tab` key to access different sections of the menu (search and footer) instead of closing the menu. Lmk if it would be better to use MenuContainer again and stick with the menu components.
- Adds demo pages to context selector, application launcher, and options menu docs.
- Adds favorites demo

patternfly-build commented:
Preview: https://patternfly-react-pr-9014.surge.sh

mcarrano commented:
The preview is broken (no left nav). Needs to be rebased?

nicolethoen reviewed (commented):
I think the new demos - like the Options Menu demo and App launcher demo needs a sentence at the top to explain that, since the old version is deprecated, the component can still be built using the new Dropdown/Select (as demonstrated in the demo)

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/ActionsMenuDemo.tsx:
```diff
@@ -1,40 +1,22 @@
 import BellIcon from '@patternfly/react-icons/dist/esm/icons/bell-icon';
 
-export const ComposableActionsMenu: React.FunctionComponent = () => {
+export const ActionsMenuDemo: React.FunctionComponent = () => {
```
Similar to my above comment, it seems like this demo should be using Select instead as well based on the functionality.

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/ApplicationLauncherDemo.tsx:
```diff
@@ -1,50 +1,30 @@
 
   const onToggleClick = (ev: React.MouseEvent) => {
-    ev.stopPropagation(); // Stop handleClickOutside from handling
+    ev.stopPropagation();
```
Should be able to remove this stopPropagation as well

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/ContextSelectorDemo.tsx:
```diff
@@ -59,46 +57,8 @@ export const ComposableContextSelector: React.FunctionComponent = () => {
-  }, [isOpen, menuRef]);
-
   const onToggleClick = (ev: React.MouseEvent) => {
     ev.stopPropagation(); // Stop handleClickOutside from handling
```
Can remove this stopPropagation

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/DateSelectDemo.tsx:
```diff
@@ -0,0 +1,88 @@
+        </MenuToggle>
+      )}
+      ref={menuRef}
+      onSelect={(_ev, itemId) => setSelected(itemId as number)}
```
Can we also close the menu when a selection occurs? That would follow how single select Select examples work

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/DrilldownMenuDemo.tsx:
```diff
@@ -27,30 +27,6 @@ export const ComposableDrilldownMenu: React.FunctionComponent = () => {
-  }, [isOpen, menuRef]);
-
   const onToggleClick = (ev: React.MouseEvent) => {
     ev.stopPropagation(); // Stop handleClickOutside from handling
     setTimeout(() => {
```
Since MenuContainer handles focusing the first element, we should be able to remove the stopPropagation and the setTimeout and just have this function handle updating state.

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/FavoritesDemo.tsx:
```diff
@@ -0,0 +1,127 @@
+  const [favorites, setFavorites] = React.useState<string[]>([]);
+  const menuRef = React.useRef<HTMLDivElement>(null);
+
+  const onToggleClick = (ev: React.MouseEvent) => {
+    ev.stopPropagation();
+    setTimeout(() => {
+      if (menuRef.current) {
+        const firstElement = menuRef.current.querySelector('li > button:not(:disabled)');
+        firstElement && (firstElement as HTMLElement).focus();
+      }
+    }, 0);
+    setIsOpen(!isOpen);
+  };
```
```suggestion
  const onToggleClick = (ev: React.MouseEvent) => {
    setIsOpen(!isOpen);
  };
```

Similar to above comments

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/ActionsMenuDemo.tsx:
```diff
@@ -48,7 +30,7 @@ export const ComposableActionsMenu: React.FunctionComponent = () => {
 
   const onToggleClick = (ev: React.MouseEvent) => {
-    ev.stopPropagation(); // Stop handleClickOutside from handling
+    ev.stopPropagation();
     setTimeout(() => {
```
We can remove the stopPropagation and the setTimeout since whichever component (Dropdown, Select) should handle focus management already

thatblindgeye commented on packages/react-core/src/demos/ComposableMenu/examples/DateSelectDemo.tsx:
```diff
@@ -0,0 +1,88 @@
+  const [selected, setSelected] = React.useState<number>(0);
+  const menuRef = React.useRef<HTMLDivElement>();
+
+  const onToggleClick = (ev: React.MouseEvent) => {
+    ev.stopPropagation(); // Stop handleClickOutside from handling
+    setTimeout(() => {
+      if (menuRef.current) {
+        const firstElement = menuRef.current.querySelector('li > button:not(:disabled)');
+        firstElement && (firstElement as HTMLElement).focus();
+      }
+    }, 0);
+    setIsOpen(!isOpen);
+  };
```
Similar to above comment

```suggestion
  const onToggleClick = (ev: React.MouseEvent) => {
    setIsOpen(!isOpen);
  };
```

thatblindgeye reviewed (changes requested):
More of a nit, but if we want to move away from using "composable" in example names, we could update the directory from `demos/ComposableMenu` to `demos/CustomMenu`.

mcarrano commented on packages/react-core/src/demos/ComposableMenu/OptionsMenu.md:
```diff
@@ -0,0 +1,16 @@
+
+import SearchIcon from '@patternfly/react-icons/dist/esm/icons/search-icon';
+
+As the `OptionsMenu` component is now deprecated, an options menu may now be built using the new suite of `Menu` components. This is showcased below in the example using the new `Dropdpown` component built off of `Menu`.
```
I agree with that. It's really a select menu that does not propagate the selection value to the toggle.

kmcfaul replied:
Will update the example to use Select. I think our initial OptionsMenu was built off of Dropdown so that's why I started with that.

mcarrano reviewed (commented):
This looks great. I have nothing to add beyond the issues already raised by @thatblindgeye .

kmcfaul commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -32,6 +32,8 @@ export interface DropdownProps extends MenuProps, OUIAProps {
   zIndex?: number;
   /** Additional properties to pass to the Popper */
   popperProps?: Partial<PopperProps>;
+  /** Keys that trigger onOpenChange, defaults to tab and escape. */
```
I lean towards the flexibility personally, but if we rather limit what a user can press to trigger `onOpenChange` I can update the prop. I do vote to keep the prop named as is to allow for more or less keys as needed in the future (maybe as an object with `Tab`/`Escape` as bool fields) rather than have a separate prop just for `tab` though. LMK

nicolethoen commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -0,0 +1,100 @@
+  }, [isOpen, menuRef, onOpenChange, onOpenChangeKeys, toggleRef]);
+
+  return (
+    <div ref={containerRef} {...props}>
```
As long as it works no matter which `appendTo` value is used.

nicolethoen commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -0,0 +1,100 @@
+
+export interface MenuContainerProps {
+  /** Menu to be rendered */
+  menu: React.ReactElement<any, string | React.JSXElementConstructor<any>>;
```
I could see it going either way, but I think the name 'menuContainer' makes me lean toward advocating for passing the menu in via `children`.

kmcfaul replied:
After tinkering with it a bit, `children` may not work very well because of the type that Popper is expecting for the popper menu. It works out for Select/Dropdown because they create the menu internally, but a passed children in MenuContainer would be a ReactNode instead of the JSXConstructor that Popper needs and errors out.

thatblindgeye commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -0,0 +1,94 @@
+  toggleRef: React.RefObject<any>;
+  /** Flag to indicate if menu is opened.*/
+  isOpen: boolean;
+  /** Callback to change the open state of the menu.
+   * Triggered by clicking outside of the menu, or by pressing either tab or escape (or keys specified in onOpenChangeKeys). */
+  onOpenChange?: (isOpen: boolean) => void;
+  /** @beta Keys that trigger onOpenChange, defaults to tab and escape. */
+  onOpenChangeKeys?: string[];
```
Just a nit: these descriptions could be updated to match the descriptions in Dropdown

nicolethoen reviewed (commented):
Can we add Props tables for components used to build these demos to this page? In particular i'm interested in seeing MenuContainer documented, but since there are not too many components used in these demos, a few other props tables may be helpful, too?

kmcfaul commented:
@nicolethoen Do you also want to add prop tables to the custom menus page? I added them to the specific demos for the deprecated components.

nicolethoen commented:
@kmcfaul i'd vote for the `MenuContainer` at least to be documented.
Do any of our `Menu` examples include a `MenuContainer`? I'd only include that props table on the `Menu` docs if they use it in the examples.

edonehoo commented on packages/react-core/src/demos/CustomMenus/CustomMenus.md:
```diff
@@ -27,78 +27,56 @@ import styles from '@patternfly/react-styles/css/components/Menu/menu';
 ## Demos
 
-Custom menus can be constructed using a composable approach by combining the [Menu](/components/menus/menu) and [Menu toggle](/components/menus/menu-toggle) components in unique ways. Composable menus currently require consumer keyboard handling and use of our undocumented [popper.js](https://popper.js.org/) wrapper component called Popper. We understand this is inconvientent boilerplate and thes ...
+Custom menus can be constructed using a composable approach by combining the [Menu](/components/menus/menu) and [Menu toggle](/components/menus/menu-toggle) components in unique ways. [Dropdown](/components/menus/dropdown), [Select](/components/menus/select), or [MenuContainer](/components/menus/menu#menucontainer) may be used in combination with menu components to handle basic keyboard inputs, o ...
```
```suggestion
Custom menus can be constructed using a composable approach by combining the [menu](/components/menus/menu) and [menu toggle](/components/menus/menu-toggle) components in unique ways. To handle basic keyboard inputs, [dropdown](/components/menus/dropdown), [select](/components/menus/select), or [`<MenuContainer>`](/components/menus/menu#menucontainer) components may be used in combination with menu components. Additionally, menu components may be connected to each other manually through our undocumented internal [popper.js](https://popper.js.org/) wrapper component called Popper.
```
Is there anything else/more PF specific to link to for Popper? Or is the popper.js site enough for users to be able to piece together that approach?

kmcfaul replied:
I'm not sure there's a more specific link we can use, short of linking directly to our github code maybe?

edonehoo commented on packages/react-core/src/demos/CustomMenus/CustomMenus.md:
```diff
@@ -27,78 +27,56 @@ import styles from '@patternfly/react-styles/css/components/Menu/menu';
-### Composable tree view menu
+### Tree view menu
 
 When rendering a menu-like element that does not contain MenuItem components, [Panel](/components/panel) allows more flexible control and customization.
```
```suggestion
When rendering a menu-like element that does not contain `<MenuItem>` components, [panel](/components/panel) allows more flexible control and customization.
```

edonehoo commented on packages/react-core/src/demos/CustomMenus/ApplicationLauncher.md:
```diff
@@ -0,0 +1,30 @@
+import ThIcon from '@patternfly/react-icons/dist/esm/icons/th-icon';
+import pfIcon from './examples/pf-logo-small.svg';
+
+As the `ApplicationLauncher` component is now deprecated, an application launcher may now be built using the new suite of `Menu` components. This is showcased below in the example using the new `Dropdpown` component built off of `Menu`.
```
```suggestion
As the application launcher component is now deprecated, an application launcher may now be built using the new suite of menu components. This is showcased in the following example, which uses the new dropdown component that is built off of menu.
```

edonehoo commented on packages/react-core/src/demos/CustomMenus/CustomMenus.md:
```diff
@@ -27,78 +27,56 @@ import styles from '@patternfly/react-styles/css/components/Menu/menu';
-### Composable flyout
+### Flyout menu
 
 The flyout will automatically position to the left or top if it would otherwise go outside the window. The menu must be placed in a container outside the main content like Popper, [Popover](/components/popover) or [Tooltip](/components/tooltip) since it may go over the side nav.
```
```suggestion
The flyout will automatically position to the left or top if it would otherwise go outside the window. The menu must be placed in a container outside the main content like Popper, [popover](/components/popover) or [tooltip](/components/tooltip), since it may go over the side nav.
```

edonehoo commented on packages/react-core/src/demos/CustomMenus/OptionsMenu.md:
```diff
@@ -0,0 +1,17 @@
+
+import SearchIcon from '@patternfly/react-icons/dist/esm/icons/search-icon';
+
+As the `OptionsMenu` component is now deprecated, an options menu may now be built using the new suite of `Menu` components. This is showcased below in the example using the new `Select` component built off of `Menu`.
```
```suggestion
As the `<OptionsMenu>` component is now deprecated, an options menu may now be built using the new suite of menu components. This is showcased in the following example, which uses the new select component that is built off of menu.
```

kmcfaul commented:
@nicolethoen No just the custom menus page uses the new container (and I've included the props table there).

mcarrano reviewed (commented):
This looks good to me. Only have one question. I see that we've added React Demo tabs for the Context selector and Application launcher components (as intended), but still also have those demos on the Custom menus page. Did we intend to duplicate these on the Custom menu page? Can't remember our earlier convo about this. What are your thoughts @nicolethoen ?

edonehoo commented on packages/react-core/src/demos/CustomMenus/ContextSelector.md:
```diff
@@ -0,0 +1,30 @@
+
+import SearchIcon from '@patternfly/react-icons/dist/esm/icons/search-icon';
+
+As the context selector component is now deprecated, a context selector may now be built using the new suite of menu components. This is showcased in the following example, which uses the new dropdown component that is built off of menu.
```
```suggestion
As the context selector component is now deprecated, a context selector may now be built using the new suite of menu components. This is showcased in the following demo, which uses the new [dropdown](/components/menus/dropdown) component that is built off of menu.
```

edonehoo commented on packages/react-core/src/demos/CustomMenus/OptionsMenu.md:
```diff
@@ -0,0 +1,17 @@
+
+import SearchIcon from '@patternfly/react-icons/dist/esm/icons/search-icon';
+
+As the `<OptionsMenu>` component is now deprecated, an options menu may now be built using the new suite of menu components. This is showcased in the following example, which uses the new select component that is built off of menu.
```
```suggestion
As the `<OptionsMenu>` component is now deprecated, an options menu may now be built using the new suite of menu components. This is showcased in the following demo, which uses the new [select](/components/menus/select) component that is built off of menu.
```

kmcfaul commented:
@mcarrano @nicolethoen I believe we landed on having the app launcher/context selector/options menu in both places. The code is only in one location it's just referenced in two places. I can remove them from the custom menus if we want though.

mcarrano commented:
> I believe we landed on having the app launcher/context selector/options menu in both places. The code is only in one location it's just referenced in two places. I can remove them from the custom menus if we want though.

I'm OK to leave it, but just wanted to raise the question.

tlabaj reviewed (changes requested):
If we are going to have this menu container, shouldn't our Select and dropdown use it?

kmcfaul commented on packages/react-core/src/components/Dropdown/Dropdown.tsx:
```diff
@@ -54,6 +54,8 @@ export interface DropdownProps extends MenuProps, OUIAProps {
   /** Additional properties to pass to the Popper */
   popperProps?: DropdownPopperProps;
+  /** @beta Keys that trigger onOpenChange, defaults to tab and escape. Escape should always be included in the array, while Tab may be omitted if the menu contains non-menu items that are focusable. */
+  onOpenChangeKeys?: string[];
```
Moved the prop upwards. The description does recommend including Escape. I think if a user wants to prevent a key from closing I think they should have the flexibility.

kmcfaul commented on packages/react-core/src/components/Menu/examples/Menu.md:
```diff
@@ -3,7 +3,18 @@ id: Menu
+    'MenuSearchInput',
+    'MenuGroup',
+    'MenuContainer'
+  ]
```
Added an interface & added to docs.

kmcfaul commented:
> If we are going to have this menu container, shouldn't our Select and dropdown use it?

We can refactor them, but as it's internal maybe this could be a follow up?

The benefit of having separate implementations is if we want to differentiate how a Select vs a Dropdown vs a generic Menu behaves by default on the other hand.

tlabaj reviewed (commented):
looks good, I just wonder if the Menu container should be Beta. @mcarrano  thoughts?

nicolethoen commented:
> looks good, I just wonder if the Menu container should be Beta. @mcarrano thoughts?

Generally, since we will be doing more frequent major releases, I don't feel the need to mark everything beta - unless it's a completely optional enhancement or experimental. I've been made aware that Extensions cannot use beta components at all, because they cannot introduce breaking changes - and have come to realize that products are generally very hesitant to adopt any beta solutions. If we want people to use this MenuContent as they build out custom Select implementations, because it _is_ our recommended solution, i'd advocate for promoting it.

I understand any hesitation though.

tlabaj commented:
> Generally, since we will be doing more frequent major releases, I don't feel the need to mark everything beta - unless it's a completely optional enhancement or experimental. I've been made aware that Extensions cannot use beta components at all, because they cannot introduce breaking changes - and have come to realize that products are generally very hesitant to adopt any beta solutions. If we want people to use this MenuContent as they build out custom Select implementations, because it _is_ our recommended solution, i'd advocate for promoting it.
> 
> I understand any hesitation though.

I think this is a discussion we should have as a team.  I will add it to the agenda.

nicolethoen commented:
We can also add/remove a beta tag later and not hold this PR up.

kmcfaul commented:
Rebased with the select PR merge, and re added a beta tag to the drilldown example that I think I removed in error during a previous rebase (just double checked the promotion PR and its still in beta from that).

tlabaj commented on packages/react-core/src/components/Menu/MenuContainer.tsx:
```diff
@@ -0,0 +1,108 @@
+  /** Callback to change the open state of the menu.
+   * Triggered by clicking outside of the menu, or by pressing any keys specificed in onOpenChangeKeys. */
+  onOpenChange?: (isOpen: boolean) => void;
+  /** @beta Keys that trigger onOpenChange, defaults to tab and escape. It is highly recommended to include Escape in the array, while Tab may be omitted if the menu contains non-menu items that are focusable. */
```
Can we mark this whole component Beta for now and remove this tag on the prop.

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-code-editor@5.0.0-alpha.111
 - @patternfly/react-core@5.0.0-alpha.110
 - @patternfly/react-docs@6.0.0-alpha.118
 - demo-app-ts@5.0.0-alpha.94
 - @patternfly/react-table@5.0.0-alpha.112

Thanks for your contribution! :tada:
